esp/services/ws_codesign/ws_codesignService.cpp

```diff
@@ -0,0 +1,92 @@
+/*##############################################################################
+
+    HPCC SYSTEMS software Copyright (C) 2019 HPCC Systems®.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+#include "ws_codesignService.hpp"
+#include "jutil.hpp"
+
+Cws_codesignEx::Cws_codesignEx()
+{
+}
+
+Cws_codesignEx::~Cws_codesignEx()
+{
+}
+
+void Cws_codesignEx::init(IPropertyTree *cfg, const char *process, const char *service)
+{
+    if(cfg == nullptr)
+        throw MakeStringException(-1, "Cannot initialize Cws_codesignEx, cfg is NULL");
+
+    StringBuffer xpath;
+    xpath.appendf("Software/EspProcess[@name=\"%s\"]/EspService[@name=\"%s\"]", process, service);
+    m_serviceCfg.setown(cfg->getPropTree(xpath.str()));
+}
+
+bool Cws_codesignEx::onSign(IEspContext &context, IEspSignRequest &req, IEspSignResponse &resp)
+{
+    resp.setRetCode(-1);
+
+    const char* keyid = req.getKeyID();
+    const char* text = req.getText();
+    if (!keyid || !*keyid || !text || !*text)
+    {
+        resp.setErrMsg("Please provide both KeyID and Text");
+        return false;
+    }
+    int keylen = strlen(keyid);
+    if(keylen != 16 && keylen != 40)
+    {
+        resp.setErrMsg("Invalid KeyID");
+        return false;
+    }
+
+    StringBuffer cmd, output, errmsg;
+
+    cmd.appendf("gpg --list-keys %s", keyid);
+    int ret = runExternalCommand(output, errmsg, cmd.str(), "");
```

Comment:
I understand there are a number of libraries that should allow the functionality to be accessed directly, rather than access via an external command.  Here's one that I found after a quick google:https://www.gnupg.org/documentation/manuals/gpgme/.  There may be better ones out there.

Reply:
Right I saw the gpgme lib as well. But since we're probably going to re-implement this service with PKI when that is in place, I don't think it's worth it to add the gpgme dependency at this moment.